common/concurrency/tokio_task.rs

```diff
@@ -0,0 +1,399 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::future::Future;
+
+use tokio::{
+    sync::{oneshot, watch::Receiver},
+    task::JoinHandle,
+    time::{self, Duration},
+};
+use tokio_util::task::TaskTracker;
+
+/// Manages background tasks that should finish gracefully on shutdown.
+///
+/// # Design
+/// - A cloneable [`TokioTaskSpawner`] is handed to lower-level components to spawn tasks.
+/// - The tracker listens to a global `watch::Receiver<()>` shutdown signal.
+/// - When shutdown is observed, the tracker:
+///   1. `close()`s to stop accepting new tasks;
+///   2. `wait()`s until all tracked tasks finish.
+///
+/// # Guarantees
+/// - Tasks spawned through the spawner are kept alive by the Tokio runtime as usual.
+/// - **To guarantee** that long-running tasks (e.g., intervals) complete their final
+///   work on shutdown (like “act_on_shutdown”), the owner **must call** [`TokioTaskTracker::join`]
+///   before letting the runtime drop.
+///
+/// This avoids blocking in `Drop` (which can deadlock a Tokio worker) while still
+/// giving you a single place to wait for all background work to finish.
+/// For more information, see https://tokio.rs/tokio/topics/shutdown.
+#[derive(Debug)]
+pub struct TokioTaskTracker {
+    tracker: TaskTracker,
+    shutdown_receiver: Receiver<()>,
+    done_receiver: oneshot::Receiver<()>,
```

Comment:
I'd prefer if the Receivers were disambiguated here
```suggestion
    shutdown_receiver: watch::Receiver<()>,
    done_receiver: oneshot::Receiver<()>,
```